### PR TITLE
Add function for computing HiPS order

### PR DESCRIPTION
**Pull request by adl1995**
This PR addresses issue #47.

### CODE REVIEW

**Comment by tboch** [comment]
@adl1995 it would be better if you could add a test asserting a few typical cases for this function.

**Comment by adl1995** [comment]
@tboch I just added the test case.

**Comment by cdeil** [comment]
@adl1995 - This test should not use `@remote_data`. OK to load something from `hips-extra` if you want.

Also -- this function should have at least 3 test cases (probably use `pytest.mark.parametrize`, for the cases where the min and max order of tiles availalbe is used, and something in between.

You can simply change the geometry fov (zoomin in and out) to trigger the three cases.

**Comment by cdeil** [comment]
I wanted to understand what this function does just now, because it's a core piece of computation in our package.

This is as far as got:
```
def get_hips_order(geometry: WCSGeometry, hips_survey: HipsSurveyProperties) -> int:
    """Compute the tile order suited for the given geometry and hips_survey"""
    # Sky image angular resolution (pixel size in deg)
    resolution = np.min(proj_plane_pixel_scales(geometry.wcs))
    # Compute HiPS order so that tile pixel resolution will be as good as sky image resolution
    desired_order = _get_hips_order_for_resolution(hips_survey.tile_width, resolution)
    # Return the desired order, or the highest resolution available
    # Note that HiPS never has resolution less than 3,
    # and that limit is handled in _get_hips_order_for_resolution
    return np.min([desired_order, hips_survey.hips_order])


def _get_hips_order_for_resolution(tile_width, resolution):
    """TODO document a bit what this function does."""
    tile_order = np.log2(tile_width)
    full_sphere_area = 4 * np.pi * np.square(180 / np.pi)
    for candidate_tile_order in range(3, 29 + 1):
        tile_resolution = np.sqrt(full_sphere_area / 12 / 4 ** (candidate_tile_order + tile_order))

        if tile_resolution <= resolution:
            break

    return candidate_tile_order
```

The code does the same as what Thomas did, but it's (I hope) a bit easier to understand what the code does in that form. My suggestion would be to put it like that in the package.

@adl1995 or @tboch - Can you write two or three lines of comments that explain how `_get_hips_order_for_resolution` works?

I think it would be good to have this well-documented, just in case there's ever an issue with the algorithm in the future, and we or others will have to revisit it.

And @adl1995 - The most important thing is to have tests that establish the current behaviour / output of that function. Please add a parametrized test for this.

**Comment by adl1995** [comment]
@cdeil I have updated the function and test cases, however, I can't use `fov` until #49 gets merged.

**Comment by cdeil** [comment]
@adl1995 - Next steps here:

- rebase this PR on latest master
- add the HiPS order computation as the first line in `make_sky_image`
- make the high-level docs example work again (maybe it just will if you add the order computation)
- the parametrized tests you've added here don't test the important things of the function you're adding. Instead you can use one HipS, but parametrize the `geometry` to have different pixel sizes, and then for smaller sky image pixels, a higher HiPS order should be computed by `get_order`. Is it clear what the goal is here with testing this function? If no, please ask!

**Comment by cdeil** [comment]
Please also add a parametrized test (in addition to the one you have) that calls `_get_hips_order_for_resolution` directly for a few tile widths and resolutions, and asserts on the returned order.

**Comment by adl1995** [comment]
@cdeil I made another commit updating the test cases.

File: hips/draw/tests/test_simple.py [hunk 1]
```diff
@@ -53,5 +55,34 @@ def test_make_sky_image():
     assert_allclose(data[200, 994], 3717.10091363)
     assert_allclose(data[200, 995], 3402.55292158)
 
+get_hips_order_pars = [
```
Comment (cdeil) [inline hips/draw/tests/test_simple.py]: To me, `get_hips_order_pars` sounds like a function. This is just a list, so suggest to rename to `hips_order_pars`.

File: hips/draw/tests/test_simple.py [hunk 2]
```diff
@@ -53,5 +55,34 @@ def test_make_sky_image():
...
+    dict(order=7, width=2000, height=1000, fov="3 deg"),
```
Comment (cdeil) [inline hips/draw/tests/test_simple.py]: To test the functionality here you can simply use one fixed `width` and `height` (e.g. 200 and 100). All you need to vary and list here is some parameter that changes the image pixel size like e.g. `fov`.

File: hips/draw/tests/test_simple.py [hunk 3]
```diff
@@ -53,5 +55,34 @@ def test_make_sky_image():
...
+    full_path = get_hips_extra_file('datasets/samples/2MASS6XH/properties')
+    hips_survey = HipsSurveyProperties.read(filename=full_path)
+    geometry = WCSGeometry.create_simple(
+        skydir=SkyCoord(0, 0, unit='deg', frame='icrs'),
+        width=pars['width'], height=pars['height'], fov=pars['fov'],
+        coordsys='icrs', projection='AIT'
+    )
+    assert get_hips_order(geometry, hips_survey) == pars['order']
+
+get_hips_order_for_resolution_pars = [
```
Comment (cdeil) [inline hips/draw/tests/test_simple.py]: Same here: rename `get_hips_order_for_resolution_pars` -> `hips_order_for_resolution_pars` or something else that doesn't sound like a function (because it is a list).

File: hips/draw/tests/test_simple.py [hunk 4]
```diff
@@ -53,5 +55,34 @@ def test_make_sky_image():
...
+    dict(tile_width=512, resolution=12.32, order=3),
+    dict(tile_width=256, resolution=160.22, order=3),
```
Comment (cdeil) [inline hips/draw/tests/test_simple.py]: You should choose cases that lead to different `orders` and test the extremes, i.e. corner / edge cases of the algorithm in that function.
At the moment you only chose values that will result in `order=3`.

File: hips/draw/simple.py [hunk 5]
```diff
@@ -116,6 +117,32 @@ def fetch_tiles(healpix_pixel_indices: np.ndarray, order: int, hips_survey: Hips
         yield tile
 
 
+def get_hips_order(geometry: WCSGeometry, hips_survey: HipsSurveyProperties) -> int:
+    """Compute the tile order suited for the given geometry and hips_survey"""
```
Comment (cdeil) [inline hips/draw/simple.py]: Add `Parameters` and `Returns` sections documenting the inputs and outputs.
Maybe even an `Examples` section with one working example so that users can see how it works in the docs?

**Comment by adl1995** [comment]
@cdeil The high-level docs example is already working. I tried it locally.

**Comment by cdeil** [comment]
@adl1995 - Are you sure? For me it's broken in master (hangs indefinitely trying to fetch a ton of tiles). Maybe you don't see it here and have to rebase on master to see it.

The first step that computes the HipS order to use for drawing to use is missing, so it's expected that it doesn't work for most cases, no?

My suggestion would be that you rebase here to see the issue, and then fix it in this PR.

**Comment by adl1995** [comment]
@cdeil I tested it in my currently branch `get_order`, in which the first step of `make_sky_image` is computing the HiPS order (https://github.com/hipspy/hips/pull/58/files#diff-166822ec0a0c3100c6f9ffdd92c4b76eR163), so it works.

**Comment by cdeil** [comment]
@adl1995 - Apologies. I missed the `order = get_hips_order(geometry, hips_survey)` line. That's perfect.

Probably this is ready to be merged then.

Do you understand why drawing tests still work without modifications?
Probably somewhat by chance, or because we don't have any drawing test that relies on the `order = get_hips_order(geometry, hips_survey)` step?
Can you add one? Do you want to do it here or in a follow-up PR?

**Comment by adl1995** [comment]
@cdeil I made some further changes adding a doctest example and updating the test case. The `draw_sky_image` function does not rely on the `order` parameter, so I'm not sure about your previous comment.

**Comment by cdeil** [comment]
This PR fixes a very important bug in the `make_sky_image` function, yet the test still passes unchanged: https://github.com/hipspy/hips/blob/master/hips/draw/tests/test_simple.py#L46

This clearly shows that the previous test wasn't very good and you should update it to something that fails if you comment out the `get_hips_order(geometry, hips_survey)` line, and succeeds after adding that fix in `make_sky_image`, no?

**Comment by adl1995** [comment]
@cdeil I haven't been able to find a survey which does not work without the `get_hips_order` function. I tried several from the `hips-extra` repo which support FITS tiles.

**Comment by cdeil** [comment]
If you go to the master branch and run the high-level docs example, and add a few print statements in https://github.com/hipspy/hips/blob/master/hips/draw/simple.py#L138 to show what `hips_survey.hips_order` is and `len(healpix_pixel_indices)`, what do you get?

I expect that you'll find that this hips_order is the maximum availabe tile resolution of the Hips and you'll get 1000s or more of tiles, i.e. draw at the wrong order.
Please try to find such a case and let me know if you don't succeed.

**Comment by adl1995** [comment]
@cdeil I just tried this and the list of tiles is 392. Should I use the same example in the test case?

**Comment by adl1995** [comment]
@cdeil I made some changes. Please review.

**Comment by adl1995** [comment]
@cdeil Can you please merge this pull request (if no changes are required)? I want to start refactoring the `SimpleTilePainter` class.

**Comment by cdeil** [comment]
OK, merging now. Thanks!
